src/main/kotlin/com/exactpro/th2/common/metrics/MetricsUtils.kt

```diff
@@ -59,6 +65,64 @@ fun incrementTotalMetrics(
     }
 }
 
+data class SessionStats(
+    var sequence: Long = 0L,
+    var groups: Int = 0,
+    var messages: Int = 0,
+    var rawMessages: Int = 0,
+)
+
+fun incrementTotalMetrics(
+    batch: MessageGroupBatch,
+    th2Pin: String,
+    messageCounter: Counter,
+    groupCounter: Counter,
+    gauge: Gauge,
+) {
+    val incomingStatsBySession = mutableMapOf<String, SessionStats>()
+    val outgoingStatsBySession = mutableMapOf<String, SessionStats>()
+
+    for (group in batch.groupsList) {
+        val messages = group.messagesList
+
+        if (messages.isEmpty()) continue
+
+        val firstMessage = messages[0]
+        val id = if (firstMessage.hasMessage()) firstMessage.message.metadata.id else firstMessage.rawMessage.metadata.id
+        val sessionAlias = id.connectionId.sessionAlias
+
+        val stats = when (id.direction) {
+            FIRST -> incomingStatsBySession.getOrPut(sessionAlias, ::SessionStats)
+            else -> outgoingStatsBySession.getOrPut(sessionAlias, ::SessionStats)
+        }
+
+        stats.groups++
+
+        messages.forEach {
+            when {
+                it.hasMessage() -> stats.messages++
+                else -> stats.rawMessages++
+            }
+        }
+
+        stats.sequence = max(stats.sequence, id.sequence)
+    }
+
+    incomingStatsBySession.forEach { (alias, stats) ->
+        gauge.labels(th2Pin, alias, FIRST.name).set(stats.sequence.toDouble())
+        groupCounter.labels(th2Pin, alias, FIRST.name).inc(stats.groups.toDouble())
+        messageCounter.labels(th2Pin, alias, FIRST.name, MESSAGE.name).inc(stats.messages.toDouble())
+        messageCounter.labels(th2Pin, alias, FIRST.name, RAW_MESSAGE.name).inc(stats.rawMessages.toDouble())
+    }
+
+    outgoingStatsBySession.forEach { (alias, stats) ->
+        gauge.labels(th2Pin, alias, SECOND.name).set(stats.sequence.toDouble())
+        groupCounter.labels(th2Pin, alias, SECOND.name).inc(stats.groups.toDouble())
+        messageCounter.labels(th2Pin, alias, SECOND.name, MESSAGE.name).inc(stats.messages.toDouble())
+        messageCounter.labels(th2Pin, alias, SECOND.name, RAW_MESSAGE.name).inc(stats.rawMessages.toDouble())
+    }
```

Comment:
Are we doing that aggregation in order to avoid multiple called of `labels` methods?

Reply:
Yes, this code is source of performance problem. Any box can receives or sends several message groups.
Maybe reduce label numbers and exclude: message name or other?

Reply:
I don't think we should reduce labels number